util/bitmap/concurrent.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bitmap
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+const (
+	segmentWidth             = 32
+	segmentWidthPower        = 5
+	bitMask           uint32 = 0x80000000
+)
+
+var bytesConcurrentBitmap = int64(int(unsafe.Sizeof(ConcurrentBitmap{})))
+
+// ConcurrentBitmap is a static-length bitmap which is thread-safe on setting.
+// It is implemented using CAS, as atomic bitwise operation is not supported by
+// golang yet. (See https://github.com/golang/go/issues/24244)
+// CAS operation is narrowed down to uint32 instead of longer types like uint64,
+// to reduce probability of racing.
+type ConcurrentBitmap struct {
+	segments []uint32
```

Comment:
The bit map may be not suitable if we must handle (`matched`, `isNull`, `unmatched`) situations in the future.

Reply:
It can use two bits of a int value.

Reply:
EMM, perhaps it's better to use uint8, it makes code simpler and faster, by just 4x memory usage.

Reply:
The CAS does not support operating the uint8.